keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,636 @@
+admit the Workload, because we only kept track of flavor assignments of type 0.
+This caused ClusterQueues in the cohort to continue
+borrowing quota, even if there were pending Workloads that would fit under the
+min Quota for their ClusterQueues.
+
+It is actually possible to limit borrowing only for the involved flavors,
+instead of restricting borrowing for all the flavors in the cohort. But we will
+leave this as a future possible optimization to improve throughput.
+
+#### Preemption
+
+For each Workload that got flavor assignments of type 1 or 2, we might need to
+preempt some admitted Workloads.
+
+The algorithm goes like follows:
+
+0. Check that `ClusterQueue.spec.waitBeforePreemptionSeconds` is satisfied.
+   The time is based on the `lastTransitionTime` of the condition `Admitted`
+   (with `status=False`). If it doesn't exist, we use the current time.
+   If the period is not satisfied, we abort preemption.
+   
+   If this is a ClusterQueue with BestEffortFIFO, the Workload is put in the
+   inadmissible staging area with a deadline equal to the remaining time to
+   fullfil `waitBeforePreemptionSeconds`.
+
+1. Check whether preemption could help.
+
+   Preemption is limited to Workloads with a priority lower than the incoming
+   Workload. To avoid unnecessary preemption calculations, we can keep a
+   priority queue with the priorities of active Workloads in the ClusterQueue
+   and the cohort. If the lowest priority in these sets is higher than or equal
+   to the priority of the incoming Workload, we can skip the preemption
+   algorithm.
+
+2. Obtain a list of victim Workloads to be preempted.
+
+  1. If the workload has flavor assignments of type 1, add the Workloads from
+     other ClusterQueues in the cohort that are over their min quota.
+  2. If the workload has flavor assigments of type 2, add the Workloads from
+     the same ClusterQueue.
+
+  When going over these sets, we filter out the Workloads that have a priority
+  greather than or equal to the priority of the incoming Workload or that are
+  not using the flavors that were selected for the incoming Workload.
+  
+  If the list is empty, abort preemption.
+
+3. Sort the Workloads using the following criteria:
+   1. Lower priority first.
+   2. Shortest running time first.
+
+4. In the order of the list, remove a Workload and check if the incoming
+   Workload fits.
+
+   This is a maximal set of Workloads that need to be removed.
+
+5. In the reverse order of the Workloads that were removed, add Workloads back
+   as long as the incoming Workload still fits.
+
+6. Preempt the Workloads.
+
+   If the workload has `.spec.onPreemption=Terminate`, add a `Finished`
+   condition with reason `Preempted`.
+
+   If the workload has `.spec.onPreemption=Requeue`, clear `.spec.admission`.
+   The Workload will be requeued by the Workload event handler.
+
+The incoming Workload will be admitted once the changes in the victim Workloads
```

Comment:
Then, change the time between the workload and the update cache where the incoming workload is placed? will it re-queue again?